File: DeepSpaceRuby/lib/Hangar.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Deepspace
  require_relative 'HangarToUI'
  require_relative 'ShieldBooster'
  require_relative 'Weapon'
  
  class Hangar
    
    attr_reader :maxElements
    attr_reader :shieldBoosters
    attr_reader :weapons
   
    
   def initialize(n)
    @maxElements=n
    @shieldBoosters = Array.new #array
    @weapons = Array.new #array
   end
     
   
      
   def self.newCopy(h)
    nuevo =  new(h.maxElements)
     for i in (0...h.shieldBoosters.size)
     nuevo.addShieldBooster(h.shieldBoosters[i])
     end
     for j in (0...h.weapons.size)
     nuevo.addWeapon(h.weapons[i])
     end
     return nuevo
   end
   
   def getUIversion
     HangarToUI.new(self)
   end
   
   # @param weapon
   # @return bool
   def addWeapon(w)
     if spaceAvailable 
       @weapons.push(w)
       return true
     else 
       puts "No hay espacio."
       return false
     end
   end
   
   #param ShieldBooster
   # @return bool
   def addShieldBooster(s)
      if spaceAvailable 
       @shieldBoosters.push(s)
       return true
     else 
       return false
     end
   end
   
   def removeShieldBooster(n)
     if @shieldBoosters != nil && n<@shieldBoosters.size
     @shieldBoosters.delete_at(n)
     end
   end
   
   def removeWeapon(n)
     if @weapons !=nil && n<@weapons.size
     @weapons.delete_at(n)
     end
   end
   
   def to_s
      "\n Elementos max: #{@maxElements}\t"
   end
   
   private
   
   def spaceAvailable
      @weapons.size+@shieldBoosters.size < @maxElements
   end
   
   
   end  
end

File: DeepSpaceJava/src/deepspace/Hangar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */


package deepspace;
import java.util.ArrayList;
/**
 *
 * @author laura
 */
public class Hangar implements Copyable<Hangar>{
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters= new ArrayList();
    private ArrayList<Weapon> weapons = new ArrayList();
    
    Hangar(int capacity){
        maxElements=capacity;
    }
    
    Hangar(Hangar h){
        maxElements=h.maxElements;
        shieldBoosters=h.shieldBoosters;
        weapons=h.weapons;
    }
    
    HangarToUI getUIversion(){
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable(){
            return (shieldBoosters.size()+weapons.size()<maxElements);
    }
    
    public boolean addWeapon(Weapon w){
        if (spaceAvailable()){
            weapons.add(w);
            return true;
        }
        else
            return false;
    }
    
    public boolean addShieldBooster(ShieldBooster s){
        if (spaceAvailable()){
            shieldBoosters.add(s);
            return true;
        }
        else
            return false;
    }
    
    public int getMaxElements(){
        return maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    } 
    public Weapon removeWeapon(int w){
        if(w>=0 && w<weapons.size())
        return weapons.remove(w);
        return null;
    }   
    
    public ShieldBooster removeShieldBooster(int s){
        if(s>=0 && s<shieldBoosters.size())
        return shieldBoosters.remove(s);
        return null;
    }

    @Override
    public String toString() {
        return "Hangar{" + "maxElements=" + maxElements + ", shieldBoosters=" + shieldBoosters + ", weapons=" + weapons + '}';
    }

    @Override
    public Hangar copy() {
       return new Hangar(this);
    }
    

    
    
    
}
